format.h

```diff
@@ -2628,16 +2639,39 @@ class BasicArrayWriter : public BasicWriter<Char> {
 typedef BasicArrayWriter<char> ArrayWriter;
 typedef BasicArrayWriter<wchar_t> WArrayWriter;
 
+template<class Elem, class Traits = std::char_traits<Elem> >
+class basic_formatbuf : public std::basic_streambuf<Elem, Traits> {
+
+  typedef typename std::basic_streambuf<Elem, Traits>::int_type int_type;
+
+  Buffer<Elem>& buffer_;
+
+public:
+    basic_formatbuf(BasicFormatter<Elem> &formatter) : buffer_(formatter.writer().buffer()) {
+      setp(buffer_.ptr_, buffer_.ptr_ + buffer_.size_, buffer_.ptr_ + buffer_.capacity_);
+  }
+
+  virtual int_type overflow(int_type _Meta = Traits::eof()) {
+    buffer_.grow(buffer_.capacity_ * 2);
+    setp(buffer_.ptr_, buffer_.ptr_ + buffer_.size_, buffer_.ptr_ + buffer_.capacity_);
+
+    return traits_type::to_int_type(*pptr());
+  }
+
+  int_type flush() {
+    buffer_.size_ = pptr() - pbase();
+    return traits_type::to_int_type(*pptr());
+  }
+};
+
 // Formats a value.
 template <typename Char, typename T>
-void format(BasicFormatter<Char> &f, const Char *&format_str, const T &value) {
-  std::basic_ostringstream<Char> os;
-  os << value;
-  std::basic_string<Char> str = os.str();
-  internal::Arg arg = internal::MakeValue<Char>(str);
-  arg.type = static_cast<internal::Arg::Type>(
-        internal::MakeValue<Char>::type(str));
-  format_str = f.format(format_str, arg);
```

Comment:
To be a bit more helpful on my suggestion, keeping the majority of your change the same, except for changing basic_format's constructor to accept a Buffer<Elem> directly:
```c++
// Formats a value.
template <typename Char, typename T>
void format(BasicFormatter<Char> &f, const Char *&format_str, const T &value) {
  internal::MemoryBuffer<Char, internal::INLINE_BUFFER_SIZE> buffer;

  basic_formatbuf<Char> format_buf(buffer);
  std::basic_ostream<Char> output(&format_buf);
  output << value;
  format_buf.flush();

  BasicStringRef<Char> str(buffer.size() > 0 ? &buffer[0] : NULL, buffer.size());
  internal::Arg arg = internal::MakeValue<Char>(str);
  arg.type = static_cast<internal::Arg::Type>(
        internal::MakeValue<Char>::type(str));
  format_str = f.format(format_str, arg);
}
```
This passes all of the tests, based on an an older commit, though.